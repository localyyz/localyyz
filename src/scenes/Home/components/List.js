import React from "react";
import {
  FlatList,
  StyleSheet,
  Text,
  View,
  TouchableWithoutFeedback
} from "react-native";

// custom
import { Colours, Styles, Sizes } from "~/src/constants";

// third party
import { observer } from "mobx-react/native";
import { withNavigation } from "react-navigation";
import PropTypes from "prop-types";
import { PropTypes as mobxPropTypes } from "mobx-react/native";

// local
import ListItem from "./ListItem";
import ListHeader from "./ListHeader";
import MoreFooter from "./MoreFooter";

@observer
export class List extends React.Component {
  static propTypes = {
    // listData input type take a look at home store
    //  fetchFeaturedProducts
    //  fetchDiscountedProducts
    // and https://github.com/mobxjs/mobx-utils#lazyobservable
    listData: mobxPropTypes.objectOrObservableObject,
    title: PropTypes.string.isRequired,
    hideHeader: PropTypes.bool,
    description: PropTypes.string,
    withMargin: PropTypes.bool,
    limit: PropTypes.number,

    // more button
    fetchPath: PropTypes.string,
    numProducts: PropTypes.number
  };

  static defaultProps = {
    title: "",
    hideHeader: false,
    description: "",
    withMargin: false
  };

  get renderMoreButton() {
    return this.props.listData.length > 0 ? (
      <MoreFooter {...this.props} />
    ) : null;
  }

  shouldComponentUpdate(nextProps) {
    return nextProps.numProducts !== this.props.numProducts;
  }

  renderItem = ({ item: product }) => {
    return <ListItem product={product} />;
  };

  gotoOnboarding = () => {
    this.props.navigation.navigate("Onboarding", {
      onFinish: this.props.navigation.goBack
    });
  };

  // TODO: this really shouldn't be abstracted here. but until
  // home is completely redone. let's leave it here for now.
  renderEmptyComponent = () => {
    return (
      <View style={styles.emptyContainer}>
        {this.props.id === "feed" ? (
          <TouchableWithoutFeedback onPress={this.gotoOnboarding}>
            <View>
              <Text style={styles.emptyText}>Nothing in your feed yet!</Text>
              <Text style={styles.emptyText}>
                Help us personalize Localyyz for you.
              </Text>
              <View
                style={[
                  Styles.RoundedButton,
                  { marginVertical: Sizes.InnerFrame }
                ]}>
                <Text style={Styles.RoundedButtonText}>Start Onboarding</Text>
              </View>
            </View>
          </TouchableWithoutFeedback>
        ) : (
          <Text style={styles.emptyText}>
            Nothing here yet. Try again later.
          </Text>
        )}
      </View>
    );
  };

  render() {
    // TODO: _position animation
    // TODO: _motion animation
    const { listData, withMargin, ...rest } = this.props;
    return (
      <View>
        {!this.props.hideHeader ? <ListHeader {...rest} /> : null}
        <View style={withMargin ? styles.listWrapper : {}}>
          <FlatList
            keyExtractor={item => `product-${item.id}`}
            renderItem={this.renderItem}
            ListFooterComponent={this.renderMoreButton}
            ListEmptyComponent={this.renderEmptyComponent}
            data={listData && listData.slice()}
            numColumns={2}/>
        </View>
      </View>
    );
  }
}

export default withNavigation(List);

const styles = StyleSheet.create({
  listWrapper: {
    padding: Sizes.InnerFrame / 2,
    backgroundColor: Colours.Foreground
  },

  emptyContainer: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    paddingHorizontal: Sizes.InnerFrame
  },

  emptyText: {
    ...Styles.Text,
    ...Styles.EmphasizedText,
    textAlign: "center"
  }
});
