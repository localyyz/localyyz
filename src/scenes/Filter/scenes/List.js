import React from "react";
import {
  View,
  StyleSheet,
  Text,
  TouchableOpacity,
  SectionList
} from "react-native";

// custom
import { Colours, Sizes, Styles, NAVBAR_HEIGHT } from "localyyz/constants";
import { ContentCoverSlider, ReactiveSpacer } from "localyyz/components";
import { capitalize } from "localyyz/helpers";
import { GA } from "localyyz/global";

// third party
import { computed } from "mobx";
import { Provider, inject, observer } from "mobx-react/native";

// constants
const SHOW_ALL_LABEL = "Show all";

export default class FilterList extends React.Component {
  constructor(props) {
    super(props);

    // data
    this.ccs = ContentCoverSlider.createStore();

    // bindings
    this.close = this.close.bind(this);
    this.onScroll = this.onScroll.bind(this);
  }

  componentDidMount() {
    this.settings.asyncFetch && this.settings.asyncFetch();
  }

  close() {
    return this.props.navigation.goBack(null);
  }

  onScroll(evt) {
    return this.sliderRef && this.sliderRef.onScroll(evt);
  }

  get sliderRef() {
    return this.refs.slider;
  }

  get title() {
    return this.settings.title;
  }

  get header() {
    return (
      <View onLayout={this.ccs.onLayout}>
        <ContentCoverSlider.Header title={this.title} />
      </View>
    );
  }

  get settings() {
    return (
      (this.props.navigation
        && this.props.navigation.state
        && this.props.navigation.state.params)
      || this.props
    );
  }

  render() {
    return (
      <Provider
        contentCoverStore={this.ccs}
        filterStore={this.settings.filterStore}>
        <ContentCoverSlider
          ref="slider"
          title={this.title}
          background={this.header}
          backColor={Colours.Text}
          backAction={this.close}>
          <Options
            {...this.settings}
            navigation={this.props.navigation}
            data={this.settings.data || []}
            onScroll={this.onScroll}/>
        </ContentCoverSlider>
      </Provider>
    );
  }
}

@inject((stores, props) => ({
  ccs: stores.contentCoverStore,
  data: stores.filterStore[props.id] || []
}))
@observer
class Options extends React.Component {
  constructor(props) {
    super(props);

    // bindings
    this.renderItem = this.renderItem.bind(this);
    this.renderSectionHeader = this.renderSectionHeader.bind(this);
    this.renderSpacer = this.renderSpacer.bind(this);
    this.renderClear = this.renderClear.bind(this);
  }

  renderItem({ item: option }) {
    return (
      <Option
        {...this.props}
        navigation={this.props.navigation}
        option={option.label || option}
        onSelect={option.onSelect || (() => this.props.setFilter(option))}/>
    );
  }

  renderSectionHeader({ section: { title } }) {
    return title ? (
      <View style={styles.wrapper}>
        <View style={styles.sectionHeader}>
          <Text style={styles.sectionHeaderLabel}>{title}</Text>
        </View>
      </View>
    ) : null;
  }

  renderSpacer() {
    return <ReactiveSpacer store={this.props.ccs} heightProp="headerHeight" />;
  }

  renderClear() {
    return this.renderItem({
      item: {
        label: SHOW_ALL_LABEL,
        onSelect: this.props.clearFilter
      }
    });
  }

  @computed
  get sections() {
    // build listing of keys and init their arrays
    let sections = Object.assign(
      {},
      ..."abcdefghijklmnopqrstuvwxyz"
        .toUpperCase()
        .split("")
        .map(letter => ({ [letter]: [] }))
    );

    // add options appropriately by first letter
    let data = this.props.data.slice();
    data.map(
      option =>
        option
        && option[0]
        && sections[option[0].toUpperCase()]
        && sections[option[0].toUpperCase()].push(option)
    );

    // and finally sort each subarray and out
    return [
      // spacer
      { data: [{}], renderItem: this.renderSpacer },

      // clear filter
      { data: [{}], renderItem: this.renderClear },

      // actual options
      ...Object.keys(sections)
        .filter(section => sections[section].length > 0)
        .map(section => ({ title: section, data: sections[section].sort() }))
    ];
  }

  render() {
    return (
      <SectionList
        {...this.props}
        sections={this.sections}
        keyExtractor={(e, i) => `filter-list-${e}-${i}`}
        renderSectionHeader={this.renderSectionHeader}
        renderItem={this.renderItem}
        showsVerticalScrollIndicator={false}
        showsHorizontalScrollIndicator={false}
        scrollEventThrottle={16}
        contentContainerStyle={styles.container}/>
    );
  }
}

class Option extends React.Component {
  constructor(props) {
    super(props);

    // bindings
    this.onSelect = this.onSelect.bind(this);
  }

  onSelect() {
    this.props.title == "Brand"
      && GA.trackEvent(
        "filter/sort",
        "filter by " + this.props.title,
        this.props.option
      );
    this.props.onSelect && this.props.onSelect();
    return this.props.navigation.goBack(null);
  }

  render() {
    return (
      <View
        style={[
          styles.wrapper,
          this.props.option == SHOW_ALL_LABEL ? styles.topWrapper : null
        ]}>
        <TouchableOpacity onPress={this.onSelect}>
          <View style={styles.option}>
            <Text style={styles.optionLabel}>
              {capitalize(this.props.option)}
            </Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    paddingBottom: Sizes.ScreenBottom + NAVBAR_HEIGHT
  },

  option: {
    paddingVertical: Sizes.InnerFrame / 2
  },

  optionLabel: {
    ...Styles.Text,
    ...Styles.Emphasized
  },

  sectionHeader: {
    paddingTop: Sizes.InnerFrame,
    paddingBottom: Sizes.InnerFrame / 2,
    marginBottom: Sizes.InnerFrame,
    borderBottomWidth: Sizes.Spacer,
    borderBottomColor: Colours.SubduedText
  },

  sectionHeaderLabel: {
    ...Styles.Text,
    ...Styles.SmallText
  },

  wrapper: {
    paddingLeft: Sizes.InnerFrame,
    backgroundColor: Colours.Foreground
  },

  topWrapper: {
    paddingVertical: Sizes.InnerFrame
  }
});
