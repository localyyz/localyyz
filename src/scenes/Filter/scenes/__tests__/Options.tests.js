import React from "react";

// custom
import Renderer from "react-test-renderer";

// local
import { Options } from "../List";
const Component = Options.wrappedComponent;

// constants
const NAME = "Filter Options";
const PROPS = {
  ccs: { headerHeight: 0 },
  data: []
};

describe(NAME, () => {
  it("should render properly", () => {
    let render = Renderer.create(<Component {...PROPS} />).getInstance();
    expect(render).toBeDefined();
  });

  it("should match snapshot", () => {
    let render = Renderer.create(<Component {...PROPS} />).toJSON();
    expect(render).toMatchSnapshot();
  });

  it("should have only two section lists (the headers) with no data", () => {
    let render = Renderer.create(<Component {...PROPS} />).getInstance();
    expect(render.sections.length).toEqual(2);
  });

  it("should have only a, b, and two headers as sections", () => {
    let render = Renderer.create(
      <Component {...PROPS} data={["a", "aa", "b", "bb", "bbb"]} />
    ).getInstance();
    expect(render.sections.length).toEqual(4);
  });
});
