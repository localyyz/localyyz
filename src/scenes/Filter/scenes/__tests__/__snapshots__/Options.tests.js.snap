// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Options should match snapshot 1`] = `
<RCTScrollView
  ListFooterComponent={
    <Component
      style={
        Object {
          "backgroundColor": "#FFFFFF",
          "padding": 7.5,
        }
      }
    />
  }
  ccs={
    Object {
      "headerHeight": 0,
    }
  }
  contentContainerStyle={
    Object {
      "paddingBottom": 48,
    }
  }
  data={
    Array [
      Object {
        "data": Array [
          Object {},
        ],
        "renderItem": [Function],
      },
      Object {
        "data": Array [
          Object {},
        ],
        "renderItem": [Function],
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={16}
  sections={
    Array [
      Object {
        "data": Array [
          Object {},
        ],
        "renderItem": [Function],
      },
      Object {
        "data": Array [
          Object {},
        ],
        "renderItem": [Function],
      },
    ]
  }
  showsHorizontalScrollIndicator={false}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={
    Array [
      0,
      3,
    ]
  }
  stickySectionHeadersEnabled={true}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "paddingTop": 0,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFFFFF",
              "paddingLeft": 15,
            },
            Object {
              "paddingTop": 15,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "paddingVertical": 7.5,
              }
            }
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "backgroundColor": "rgba(255, 255, 255, 0)",
                  "color": "#000000",
                  "fontSize": 16,
                  "fontWeight": "600",
                }
              }
            >
              Show All
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "#FFFFFF",
            "padding": 7.5,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
